src/Uno.UI/UI/Xaml/Controls/Popup/PopupPanel.Placement.cs

```diff
@@ -0,0 +1,384 @@
+using XamlWindow = Microsoft.UI.Xaml.Window;
+#else
+using WindowSizeChangedEventArgs = Windows.UI.Core.WindowSizeChangedEventArgs;
+#endif
+
+namespace Windows.UI.Xaml.Controls.Primitives;
+
+partial class PopupPanel
+{
+	// This is a base popup panel to calculate the placement near an anchor control.
+	// This class exists mostly to reuse the same logic between a Flyout and a ToolTip.
+	// This class should eventually be removed, and Uno should match WinUI's approach, where Flyout sets Popup.HorizontalOffset and VerticalOffset
+	// as well as Width and Height on FlyoutPresenter when it opens, and then allows the popup layouting to do its job.
+
+	private static readonly Dictionary<FlyoutBase.MajorPlacementMode, Memory<FlyoutBase.MajorPlacementMode>> PlacementsToTry =
+		new Dictionary<FlyoutBase.MajorPlacementMode, Memory<FlyoutBase.MajorPlacementMode>>()
+		{
+			 {FlyoutBase.MajorPlacementMode.Top, new []
+			 {
+				 FlyoutBase.MajorPlacementMode.Top,
+				 FlyoutBase.MajorPlacementMode.Bottom,
+				 FlyoutBase.MajorPlacementMode.Left,
+				 FlyoutBase.MajorPlacementMode.Right,
+				 FlyoutBase.MajorPlacementMode.Top // use preferred choice if no others fit
+			 }},
+			 {FlyoutBase.MajorPlacementMode.Bottom, new []
+			 {
+				 FlyoutBase.MajorPlacementMode.Bottom,
+				 FlyoutBase.MajorPlacementMode.Top,
+				 FlyoutBase.MajorPlacementMode.Left,
+				 FlyoutBase.MajorPlacementMode.Right,
+				 FlyoutBase.MajorPlacementMode.Bottom // use preferred choice if no others fit
+			 }},
+			 {FlyoutBase.MajorPlacementMode.Left, new []
+			 {
+				 FlyoutBase.MajorPlacementMode.Left,
+				 FlyoutBase.MajorPlacementMode.Right,
+				 FlyoutBase.MajorPlacementMode.Top,
+				 FlyoutBase.MajorPlacementMode.Bottom,
+				 FlyoutBase.MajorPlacementMode.Left // use preferred choice if no others fit
+			 }},
+			 {FlyoutBase.MajorPlacementMode.Right, new []
+			 {
+				 FlyoutBase.MajorPlacementMode.Right,
+				 FlyoutBase.MajorPlacementMode.Left,
+				 FlyoutBase.MajorPlacementMode.Top,
+				 FlyoutBase.MajorPlacementMode.Bottom,
+				 FlyoutBase.MajorPlacementMode.Right // use preferred choice if no others fit
+			 }},
+		};
+
+	/// <summary>
+	/// This value is an arbitrary value for the placement of
+	/// a popup below its anchor.
+	/// </summary>
+	protected virtual int PopupPlacementTargetMargin => 5;	
+
+	private void Window_SizeChanged(object sender, WindowSizeChangedEventArgs e)
+		=> InvalidateMeasure();
+
+	// TODO: Use this whenever popup placement is Auto
+	protected virtual PopupPlacementMode DefaultPopupPlacement { get; }
+
+	protected virtual bool FullPlacementRequested { get; }
+
+	internal virtual FlyoutBase Flyout => null;
+
+	private Size PlacementArrangeOverride(Popup popup, Size finalSize)
+	{
+		foreach (var child in Children)
+		{
+			var desiredSize = child.DesiredSize;
+			var maxSize = (child as FrameworkElement).GetMaxSize(); // UWP takes FlyoutPresenter's MaxHeight and MaxWidth into consideration, but ignores Height and Width
```

Comment:
In this line, if `child` is not `FrameworkElement`, the extension `GetMaxSize` will throw `NullReferenceException`

- If you can think of a scenario where `child` is not `FrameworkElement`, it should be tested and the code should be updated accordingly.
- Otherwise, a direct cast should be better than `as` in that if this fails, the exception will be clear about the type of `child`.

    ```suggestion
    			var maxSize = ((FrameworkElement)child).GetMaxSize(); // UWP takes FlyoutPresenter's MaxHeight and MaxWidth into consideration, but ignores Height and Width
    ```